src/raqm.c

```diff
@@ -1126,3 +1136,372 @@ _raqm_u32_to_u8_index (raqm_t   *rq,
+ * @index. If the position is outside the text, the last character is chosen as
+ * @index.
+ * 
+ * Return value:
+ * %true if the @index inside the text, %false otherwise.
+ *
+ * Since: 0.1
+ */
+bool
+raqm_position_to_index (raqm_t *rq,
+                        int x,
+                        int y,
+                        size_t *index)
+{
+  int delta_x = 0, current_x = 0;
+  if (rq == NULL)
+    return false;
+
+  if (x < 0) /* Get leftmost index */
+  {
+    if (rq->resolved_dir == RAQM_DIRECTION_RTL)
+      *index = rq->text_len;
+    else
+      *index = 0;
+    return false;
+  }
+
+  RAQM_TEST ("\n");
+
+  for (raqm_run_t *run = rq->runs; run != NULL; run = run->next)
+  {
+    size_t len;
+    hb_glyph_info_t *info;
+    hb_glyph_position_t *position;
+    len = hb_buffer_get_length (run->buffer);
+    info = hb_buffer_get_glyph_infos (run->buffer, NULL);
+    position = hb_buffer_get_glyph_positions (run->buffer, NULL);
+ 
+    for (size_t i = 0; i < len; i++)
+    {
+      delta_x = position[i].x_advance;
+      if (x < (current_x + delta_x))
+      {
+        bool before = false;
+        if (run->direction == HB_DIRECTION_LTR)
+          before = (x < current_x + (delta_x / 2));
+        else
+          before = (x > current_x + (delta_x / 2));
+
+        if (before)
+          *index = info[i].cluster;
+        else
+        {
+          uint32_t curr_cluster = info[i].cluster;
+          uint32_t next_cluster = curr_cluster;
+          if (run->direction == HB_DIRECTION_LTR)
+            for (size_t j = i + 1; j < len && next_cluster == curr_cluster; j++)
+              next_cluster = info[j].cluster;
+          else
+            for (int j = i - 1; i != 0 && j >= 0 && next_cluster == curr_cluster; j--)
+              next_cluster = info[j].cluster;
+
+          if (next_cluster == curr_cluster)
+            next_cluster = run->pos + run->len;
+
+          *index = next_cluster;
+        }
+        if (_allowed_grapheme_boundary(rq->text[*index],rq->text[*index+1]))
+        {
+          RAQM_TEST ("The start-index is %ld  at position %d \n", *index, x);
+            return true;
+        }
+
+        while (*index < (unsigned)run->pos + run->len)
+        {
+          if (_allowed_grapheme_boundary(rq->text[*index],rq->text[*index+1]))
+          {
+            *index += 1;
+            break;
+          }
+          *index += 1;
+        }
+        RAQM_TEST ("The start-index is %ld  at position %d \n", *index, x);
+        return true;
+      }
+      else
+        current_x += delta_x;
+    }
+  }
+
+  /* Get rightmost index*/
+  if (rq->resolved_dir == RAQM_DIRECTION_RTL)
+    *index = 0;
+  else
+    *index = rq->text_len;
+
+  RAQM_TEST ("The start-index is %ld  at position %d \n",
+             *index, x);
+
+  return false;
+
+}
+
+typedef enum
+{
+  CR,
+  LF,
+  CONTROL,
+  EXTEND,
+  REGIONAL_INDICATOR,
+  PREPEND,
+  SPACING_MARK,
+  HANGUL_SYLLABLE,
+  OTHER
+} GraphemeBreak;
```

Comment:
Use lowercase for the name of the enum, and use `raqm_` prefix, so it should be `raqm_grapheme_t`. For the value, it should be also something like `RAQM_GRAPHEM_CR` etc. as using very short names like `CR` or `LF` can potentially conflict with other libraries.